### DISCUSSION TITLE
Pass declared record key as parameter

### DISCUSSION THREAD

**Discussion by PedroHase** [teal-language/tl#602] [Q&A] [answered]
Hey all, I want to declare a function from a global extension where a parameter should hold the value of a record.

E.g. suppose I have following function that I want to declare:

```
MP.CreateEventTimer(event_name: string, interval_ms: number, [strategy: CallStrategy])
```

`CallStrategy` encodes a number value and is exposed by the extension through

```
MP.CallStrategy.BestEffort
MP.CallStrategy.Precise
```

How would I be able to declare `CreateEventTimer` so that I would be able to call?

```lua
MP.CreateEventTimer("testEvent", 1000, MP.CallStrategy.BestEffort)
```

<details>
<summary>

My declaration attempt: [Teal-Playground](https://teal-playground.netlify.app/?c=LS0gZGVjbGFyZSBNUApnbG9iYWwgcmVjb3JkIE1QCiAgICByZWNvcmQgQ2FsbFN0cmF0ZWd5CiAgICAgICAgQmVzdEVmZm9ydDogYW55CiAgICAgICAgUHJlY2lzZTogYW55CiAgICBlbmQKICAgIENyZWF0ZUV2ZW50VGltZXI6IGZ1bmN0aW9uKGV2ZW50TmFtZTogc3RyaW5nLCB0aW1lckluTVM6IG51bWJlciwgc3RyYXRlZ3k6IE1QLkNhbGxTdHJhdGVneSk6IG5pbAplbmQKCi0tIHRlc3QgQ3JlYXRlRXZlbnRUaW1lcgpNUC5DcmVhdGVFdmVudFRpbWVyKCJ0ZXN0RXZlbnQiLCAxMDAwLCBNUC5DYWxsU3RyYXRlZ3kuUHJlY2lzZSk%3D)

</summary>

```lua
-- declare MP
global record MP
    record CallStrategy
        BestEffort: any
        Precise: any
    end
    CreateEventTimer: function(eventName: string, timerInMS: number, strategy: MP.CallStrategy): nil
end

-- test CreateEventTimer
MP.CreateEventTimer("testEvent", 1000, MP.CallStrategy.Precise)
-- Error: argument 3: got <any type>, expected MP.CallStrategy
```
</details>

**Comment by hishamhm** [comment, marked as answer]
@PedroHase You can do something like this:

```lua
-- declare MP
global record MP
    record CallStrategyType
        -- what do call strategies have in common?
    end
    record CallStrategy
        BestEffort: CallStrategyType
        Precise: CallStrategyType
    end
    CreateEventTimer: function(eventName: string, timerInMS: number, strategy: MP.CallStrategyType): nil
end

-- test CreateEventTimer
MP.CreateEventTimer("testEvent", 1000, MP.CallStrategy.Precise)
```

**Comment by hishamhm** [comment]
With the code in the latest master (which fixes typechecking of enum keys in maps when using dot-notation), you could alternatively declare CallStrategy as a map like this:

```lua
-- declare MP
global record MP
    record CallStrategyType
        -- what do call strategies have in common?
    end
    enum CallStrategyName
        "BestEffort"
        "Precise"
    end
    CallStrategy: {CallStrategyName:CallStrategyType}
    CreateEventTimer: function(eventName: string, timerInMS: number, strategy: MP.CallStrategyType): nil
end

-- test CreateEventTimer
MP.CreateEventTimer("testEvent", 1000, MP.CallStrategy.Precise)
```

**Reply by PedroHase** [reply]
Thanks for your help @hishamhm, highly appreciated 😊 ! Is there an ETA for when the next release will drop that contains the fix (i.e. if it would be better to wait, or use your first solution)?

**Reply by hishamhm** [reply]
I think the first solution works just as well.

I want to get some feedback for a few days to see if there are any uncaught regressions in the code I pushed recently, so I'll give at least a couple of weeks until the next release. You can try the latest code now with `luarocks install tl --dev` if you're running it from LuaRocks (`luarocks install tl` will install the stable release back).
